tools/traffic_vault_migrate/postgres.go

```diff
@@ -0,0 +1,716 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	util "github.com/apache/trafficcontrol/lib/go-util"
+	_ "github.com/lib/pq"
+	"log"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+type PGConfig struct {
+	Host     string `json:"host"`
+	Port     string `json:"port"`
+	User     string `json:"user"`
+	Password string `json:"password"`
+	SSLMode  string `json:"sslmode"`
+	Database string `json:"database"`
+	Key      string `json:"aesKey"`
+	AESKey   []byte
+}
+type PGBackend struct {
+	sslKey PGSSLKeyTable
+	dnssec PGDNSSecTable
+	uri    PGURISignKeyTable
+	url    PGURLSigKeyTable
+	cfg    PGConfig
+	db     *sql.DB
+}
+
+func (pg *PGBackend) String() string {
+	data := fmt.Sprintf("PG server %v@%v:%v\n", pg.cfg.User, pg.cfg.Host, pg.cfg.Port)
+	data += fmt.Sprintf("\tSSL Keys: %v\n", len(pg.sslKey.Records))
+	data += fmt.Sprintf("\tDNSSec Keys: %v\n", len(pg.dnssec.Records))
+	data += fmt.Sprintf("\tURI Keys: %v\n", len(pg.uri.Records))
+	data += fmt.Sprintf("\tURL Keys: %v\n", len(pg.url.Records))
+	return data
+}
+func (pg *PGBackend) Name() string {
+	return "PG"
+}
+func (pg *PGBackend) ReadConfig(s string) error {
+	genericCfg, err := UnmarshalConfig(s, reflect.TypeOf(pg.cfg))
+	if err != nil {
+		return err
+	}
+
+	pg.cfg = *genericCfg.Interface().(*PGConfig)
+
+	pg.cfg.AESKey, err = base64.StdEncoding.DecodeString(pg.cfg.Key)
+	if err != nil {
+		return fmt.Errorf("unable to decode PG AESKey: %v", err)
+	}
+	return nil
+}
+func (pg *PGBackend) Insert() error {
+	if err := pg.sslKey.insertKeys(pg.db); err != nil {
+		return err
+	}
+	if err := pg.dnssec.insertKeys(pg.db); err != nil {
+		return err
+	}
+	if err := pg.url.insertKeys(pg.db); err != nil {
+		return err
+	}
+	if err := pg.uri.insertKeys(pg.db); err != nil {
+		return err
+	}
+	return nil
+}
+func (pg *PGBackend) Start() error {
+	sqlStr := fmt.Sprintf("postgres://%s:%s@%s:%s/%s?sslmode=%s", pg.cfg.User, pg.cfg.Password, pg.cfg.Host, pg.cfg.Port, pg.cfg.Database, pg.cfg.SSLMode)
+	db, err := sql.Open("postgres", sqlStr)
+	if err != nil {
+		return fmt.Errorf("Unable to start PG client: %v", err)
+	}
+
+	pg.db = db
+	pg.sslKey = PGSSLKeyTable{}
+	pg.dnssec = PGDNSSecTable{}
+	pg.url = PGURLSigKeyTable{}
+	pg.uri = PGURISignKeyTable{}
+
+	return nil
+}
+func (pg *PGBackend) ValidateKey() []string {
+	var errors []string
+	if errs := pg.sslKey.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	if errs := pg.dnssec.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	if errs := pg.uri.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	if errs := pg.url.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	return errors
+}
+func (pg *PGBackend) Stop() error {
```

Comment:
Do you think this would benefit from being an [`io.Closer`](https://pkg.go.dev/io#Closer) interface instead of using `Stop`?

Reply:
Don't see the harm, renamed it.